index.js

```diff
@@ -79,41 +86,24 @@ function arg(opts, {argv, permissive = false} = {}) {
 					}
 				}
 
-				/* eslint-disable operator-linebreak */
-				const [type, isArray] = Array.isArray(handlers[argName])
-					? [handlers[argName][0], true]
-					: [handlers[argName], false];
-				/* eslint-enable operator-linebreak */
+				const [type, isFlag] = handlers[argName];
 
-				if (!(type === Boolean || type[flagSymbol]) && ((j + 1) < separatedArguments.length)) {
+				if (!isFlag && ((j + 1) < separatedArguments.length)) {
 					throw new TypeError(`Option requires argument (but was followed by another short argument): ${originalArgName}`);
 				}
 
-				let value;
-				if (type === Boolean) {
-					value = true;
-				} else if (type[flagSymbol]) {
-					value = type(true, argName, result[argName]);
+				if (isFlag) {
+					result[argName] = type(true, argName, result[argName]);
 				} else if (argStr === undefined) {
 					if (argv.length < i + 2 || (argv[i + 1].length > 1 && argv[i + 1][0] === '-')) {
 						const extended = originalArgName === argName ? '' : ` (alias for ${argName})`;
 						throw new Error(`Option requires argument: ${originalArgName}${extended}`);
 					}
 
-					value = type(argv[i + 1], argName, result[argName]);
+					result[argName] = type(argv[i + 1], argName, result[argName]);
 					++i;
 				} else {
-					value = type(argStr, argName, result[argName]);
-				}
-
-				if (isArray) {
-					if (result[argName]) {
-						result[argName].push(value);
-					} else {
-						result[argName] = [value];
-					}
-				} else {
-					result[argName] = value;
+					result[argName] = type(argStr, argName, result[argName]);
```

Comment:
Previously this required building up `value` and then assigning it at the end, with conditional logic for handling the array type. By creating the `of(fn)` wrapper, it moves this logic into the handler itself so the logic no longer lives within the loop.